flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/metrics/MetricFetcher.java

```diff
@@ -101,127 +98,112 @@ public void update() {
 		}
 	}
 
-	private void logErrorOnFailure(Future<Object> future, final String message) {
-		future.onFailure(new OnFailure() {
-			@Override
-			public void onFailure(Throwable failure) throws Throwable {
-				LOG.debug(message, failure);
-			}
-		}, ctx);
-	}
-
 	/**
-	 * Requests a metric dump from the given actor.
+	 * Retrieves and queries the specified QueryServiceGateway.
 	 *
-	 * @param actor ActorRef to request the dump from
-     */
-	private void queryMetrics(ActorRef actor) {
-		Future<Object> metricQueryFuture = new BasicGateway(actor).ask(MetricQueryService.getCreateDump(), timeout);
-		metricQueryFuture
-			.onSuccess(new OnSuccess<Object>() {
-				@Override
-				public void onSuccess(Object result) throws Throwable {
-					addMetrics(result);
+	 * @param queryServicePath specifying the QueryServiceGateway
+	 */
+	private void retrieveAndQueryMetrics(String queryServicePath) {
+		final CompletableFuture<MetricQueryServiceGateway> jmQueryServiceGatewayFuture = queryServiceRetriever.retrieveService(queryServicePath);
+
+		jmQueryServiceGatewayFuture.whenCompleteAsync(
+			(MetricQueryServiceGateway queryServiceGateway, Throwable t) -> {
+				if (t != null) {
+					LOG.debug("Could not retrieve QueryServiceGateway.", t);
+				} else {
+					queryMetrics(queryServiceGateway);
 				}
-			}, ctx);
-		logErrorOnFailure(metricQueryFuture, "Fetching metrics failed.");
-	}
-
-	private void addMetrics(Object result) {
-		MetricDumpSerialization.MetricSerializationResult data = (MetricDumpSerialization.MetricSerializationResult) result;
-		List<MetricDump> dumpedMetrics = deserializer.deserialize(data);
-		for (MetricDump metric : dumpedMetrics) {
-			metrics.add(metric);
-		}
+			},
+			executor);
 	}
 
 	/**
-	 * Helper class that allows mocking of the answer.
-     */
-	static class BasicGateway {
-		private final ActorRef actor;
-
-		private BasicGateway(ActorRef actor) {
-			this.actor = actor;
-		}
-
-		/**
-		 * Sends a message asynchronously and returns its response. The response to the message is
-		 * returned as a future.
-		 *
-		 * @param message Message to be sent
-		 * @param timeout Timeout until the Future is completed with an AskTimeoutException
-		 * @return Future which contains the response to the sent message
-		 */
-		public Future<Object> ask(Object message, FiniteDuration timeout) {
-			return Patterns.ask(actor, message, new Timeout(timeout));
-		}
+	 * Query the metrics from the given QueryServiceGateway.
+	 *
+	 * @param queryServiceGateway to query for metrics
+	 */
+	private void queryMetrics(final MetricQueryServiceGateway queryServiceGateway) {
+		queryServiceGateway
+			.queryMetrics(timeout)
+			.whenCompleteAsync(
+				(MetricDumpSerialization.MetricSerializationResult result, Throwable t) -> {
+					if (t != null) {
+						LOG.debug("Fetching metrics failed.", t);
+					} else {
+						List<MetricDump> dumpedMetrics = deserializer.deserialize(result);
+						for (MetricDump metric : dumpedMetrics) {
```

Comment:
Yes, there's still a discussion about it, but that's not an argument against fixing it now to make it behave the way it was originally intended to.